tools/generator/templates/player/haskell/CApi.hsc.jinja2

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE ForeignFunctionInterface #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{{ stechec2_generated|haskell_comment }}
+
+module CApi where
+
+import Data.List (foldl')
+import Foreign
+import Foreign.C.Types
+import Foreign.Marshal.Array
+
+#include "api.h"
+
+class ApiStorable a where
+  type ApiStorableType a
+  toStorable :: a -> (ApiStorableType a -> IO b) -> IO b
+  unStorable :: ApiStorableType a -> IO a
+  type ApiStorableBaseType a
+  toStorableBase :: a -> (ApiStorableBaseType a -> IO b) -> IO b
+  unStorableBase :: ApiStorableBaseType a -> IO a
+
+instance ApiStorable () where
+  type ApiStorableType () = ()
+  toStorable () f = f ()
+  {-# INLINE toStorable #-}
+  unStorable _ = return ()
+  {-# INLINE unStorable #-}
+  type ApiStorableBaseType () = ()
+  toStorableBase = toStorable
+  {-# INLINE toStorableBase #-}
+  unStorableBase = unStorable
+  {-# INLINE unStorableBase #-}
+
+instance ApiStorable Int where
+  type ApiStorableType Int = CInt
+  toStorable x f = f (fromIntegral x)
+  {-# INLINE toStorable #-}
+  unStorable = return . fromIntegral
+  {-# INLINE unStorable #-}
+  type ApiStorableBaseType Int = ApiStorableType Int
+  toStorableBase = toStorable
+  {-# INLINE toStorableBase #-}
+  unStorableBase = unStorable
+  {-# INLINE unStorableBase #-}
+
+instance ApiStorable Bool where
+  type ApiStorableType Bool = CBool
+  toStorable x f = f ((\bx -> if bx then 1 else 0) x)
+  {-# INLINE toStorable #-}
+  unStorable = return . (/= 0)
+  {-# INLINE unStorable #-}
+  type ApiStorableBaseType Bool = ApiStorableType Bool
+  toStorableBase = toStorable
+  {-# INLINE toStorableBase #-}
+  unStorableBase = unStorable
+  {-# INLINE unStorableBase #-}
+
+instance ApiStorable Double where
+  type ApiStorableType Double = CDouble
+  toStorable x f = f (CDouble x)
+  {-# INLINE toStorable #-}
+  unStorable = return . (\(CDouble x) -> x)
+  {-# INLINE unStorable #-}
+  type ApiStorableBaseType Double = ApiStorableType Double
+  toStorableBase = toStorable
+  {-# INLINE toStorableBase #-}
+  unStorableBase = unStorable
+  {-# INLINE unStorableBase #-}
+
+{% for enum in game.enum %}
+
+{{ enum.enum_summary|haskell_comment }}
+data {{ enum.enum_name|capitalize }} =
+    {{ enum.enum_field|haskell_enum_fields }}
```

Comment:
Just use something like this instead of a custom filter?

```suggestion
    {% for field_name, field_comment in enum.enum_fields %}
    {% if not loop.first %}|{% else %} {% endif %} {{ field_name.capitalize() }} -- {{ field_comment }}
    {% endfor %}
```